### ISSUE TITLE
[api-extractor] Renames exports that have the same name as a global

### ISSUE THREAD

**Issue by rbuckton** [microsoft/rushstack#1830]
**Is this a feature or a bug?**
- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

API extractor renames symbols that share the same name as a global resulting in incorrect output
for doc model, api report, and dts rollup output.

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

Run api-extractor on a project that has something like `export class Node {}` or `export class Symbol {}`.

**What is the expected behavior?**

The names in the various outputs should be the actual names (including in declaration reference/uids, etc.). 

**Actual Behavior**

![image](https://user-images.githubusercontent.com/3902892/79175650-231de080-7db3-11ea-8f0a-29d817dcd058.png)

This is especially bad when combined with dts rollup as the renamed exports are not reachable.

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** api-extractor
* **Tool Version:** 7.7.13
* **Node Version:** 12.13.0
  * **Is this a LTS version?** yes 
  * **Have you tested on a LTS version?**  yes
* **OS:** Windows 10

**Comment by octogonz** [comment]
> Run api-extractor on a project that has something like `export class Node {}` or `export class Symbol {}`.

@rbuckton consider this input:

**src/index.ts**
```ts
export { f } from './utils';

/** @public */
export class Symbol {  // local Symbol hides the global
}
```

**src/utils.ts**
```ts
/** @public */
export function f(): Symbol { // returns the global Symbol (not local)
    return Symbol('hello');
}
```

API Extractor rolls them up like this:

**dist/rollup.d.ts**
```ts
/** @public */
export declare function f(): Symbol;

/** @public */
declare class Symbol_2 {
}
export { Symbol_2 as Symbol }

export { }
```

As you can see, in general API Extractor is forced to locally rename global names, because the global and local `Symbol` may coexist in the same API surface.  (For an external consumer of the API, this is no problem, since in order to consume `f()` they have the freedom/responsibility to assign names in the caller's source file.)  Since `Symbol_2` is unreachable, this slightly ugly behavior doesn't technically impact the importable types and thus seems acceptable. 

The API report follows the .d.ts rollup.  So perhaps the problem you're reporting only involves the .api.json file.

**temp/example.api.json**
```js
. . .
  {
    "kind": "Class",
    "canonicalReference": "api-extractor-lib1-test!Symbol_2:class",
    "docComment": "/**\n * @public\n */\n",
    "excerptTokens": [
      {
        "kind": "Content",
        "text": "export declare class Symbol "
      }
    ],
    "releaseTag": "Public",
    "name": "Symbol_2",
    "members": [],
    "implementsTokenRanges": []
  }
. . .
```

Here the `name` and `canonicalReference` are referencing the local name `Symbol_2`, when they should instead be referencing the (renamed) export name `Symbol`.  That does seem like a bug.

Would you agree that this is the only thing we need to fix here?

**Comment by octogonz** [comment]
(BTW this issue probably increased with API Extractor 7.7.12, which improved the detection of global variables and now discovers a lot more of these conflicts than in the past.)

**Comment by rbuckton** [comment]
That seems like an issue with dts roll up tangentially related to this one. Renaming the exported `Symbol` introduces an export that doesn't exist in the JS. The dts roll up should disambiguate the global, not the export, if possible. 

There are several ways to do this, depending on TS version:

```ts
export declare function f(): globalThis.Symbol;

export declare function f(): ReturnType<SymbolConstructor>;

// or

class Symbol_2 {}
export { Symbol_2 as Symbol }
```

Either way, it should still be `Symbol` in the documentation output.

**Comment by octogonz** [comment]
> Renaming the exported `Symbol` introduces an export that doesn't exist in the JS.

What do you mean by this?  In my above example, the export is called `Symbol`.  `Symbol_2` is not an export and essentially invisible to consumers.

**Comment by rbuckton** [comment]
I must have missed the `export { ... }` declaration when I looked at the outputs. I'm primarily concerned with the output in the documentation, however, which is still incorrect.

**Comment by octogonz** [comment]
Right, but before making the fix, I was trying to clarify what to change:

> Here the `name` and `canonicalReference` are referencing the local name `Symbol_2`, when they should instead be referencing the (renamed) export name `Symbol`. That does seem like a bug.
> 
> Would you agree that this is the only thing we need to fix here?

**Comment by rbuckton** [comment]
I think that is sufficient for my purposes, yes.

**Comment by octogonz** [comment]
Here's a preliminary sketch of a fix: https://github.com/microsoft/rushstack/pull/1832

But I wonder if we should think about the right way to represents this in the .api.json file (and API documentation).

Example input where `exportedName` is ambiguous:
```ts
enum X { }

namespace X { }

export { X as Y, X as Z };
```

**Comment by rbuckton** [comment]
I've said before, I think you should emit documentation stubs for aliases that just link to the original definition.

**Comment by octogonz** [comment]
We might be fix that in this PR.  There's a design question of which one is the alias.  For example, we could say that `Y` is the main declaration, and `Z` is the alias.  But it's an arbitrary choice (unless the local name was `Y` instead of `X`).

When my PR, the API JSON will look like this:
```js
  {
    "kind": "Enum",
    "canonicalReference": "api-extractor-test-01!Y:enum",
    "docComment": "",
    "excerptTokens": [
      {
        "kind": "Content",
        "text": "declare enum X "
      }
    ],
    "releaseTag": "Public",
    "name": "Y",
    "members": []
  },
  {
    "kind": "Namespace",
    "canonicalReference": "api-extractor-test-01!Y:namespace",
    "docComment": "",
    "excerptTokens": [
      {
        "kind": "Content",
        "text": "declare namespace X "
      }
    ],
    "releaseTag": "Public",
    "name": "Y",
    "members": []
  }
```

So maybe we add an additional record for `Z`, something like this:
```js
  {
    "kind": "ExportAlias",  // or "Alias"?
    "canonicalReference": "api-extractor-test-01!Z:alias",  // ":alias" is a new kind
     // Note that this points to the main export Y, not to the local declaration
     // whose name "X" is unreachable and meaningless:
    "targetCanonicalReference": "api-extractor-test-01!Y:alias",
    "docComment": "",
    // There are no excerptTokens, since the declaration will be
    // purely a synthetic "export {  Y as Z }".
    "excerptTokens": [ ],
    "releaseTag": "Public",
    "name": "Z"
  },
```

**Comment by octogonz** [comment]
Seems like `ExportedAlias` could also be used to represent reexports like these:

```
// "name": "A",
// "targetCanonicalReference": "other-package!A",
export { A } from `other-package`;

import { B } from `other-package-1`;

// "name": "C",
// "targetCanonicalReference": "other-package!B",
export { B as C };

// This should probably get modeled differently, e.g. "kind": "ExportStar"
export * from `another-package`;
```

**Comment by rbuckton** [comment]
```ts
enum X { }

namespace X { }

export { X as Y, X as Z };
```

My preference for this would have been to have the doc item named `X` with aliases `Y` and `Z`, and some kind of indicator of that case, i.e.,

---

### Enum X (local)
**Exported as:** [Y](#Y), [Z](#Z)
...

---

However, I understand that doesn't align with the current design of tsdoc. Aside from that, I'd say just go with the first export in source order for the default name and maybe have a doc tag that allows you to be explicit such as the [`@alias` JSDoc tag](https://jsdoc.app/tags-alias.html):
```ts
/** @alias Z */
enum X {}
namespace X {}
export { X as Y, X as Z } // Z is picked for the actual declaration.
```

**Comment by marekdedic** [comment]
I have additional use-case and I am not sure whether it falls under this issue or whether I should open a new one.

My library translates between 2 kinds of ASTs and therefore has to deal with conflicting imports. So I have imports like:

```ts
import type { Node as ProseMirrorNode } from "prosemirror-model";
import type { Node as UnistNode } from "unist";
```

However, after running this through api-extractor, I get:

```ts
import type { Node as Node_2 } from 'unist';
import type { Node as Node_3 } from 'prosemirror-model';
```

While functionally the same, it is really confusing for the users :(
